File: decompiled-src-verify/io/legado/app/model/analyzeRule/AnalyzeByXPath.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  kotlin.Metadata
 *  kotlin.collections.CollectionsKt
 *  kotlin.jvm.internal.Intrinsics
 *  kotlin.text.StringsKt
 *  org.jetbrains.annotations.NotNull
 *  org.jetbrains.annotations.Nullable
 *  org.jsoup.nodes.Document
 *  org.jsoup.nodes.Element
 *  org.jsoup.select.Elements
 *  org.seimicrawler.xpath.JXDocument
 *  org.seimicrawler.xpath.JXNode
 */
package io.legado.app.model.analyzeRule;

import io.legado.app.model.analyzeRule.RuleAnalyzer;
import io.legado.app.utils.TextUtils;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import kotlin.Metadata;
import kotlin.collections.CollectionsKt;
import kotlin.jvm.internal.Intrinsics;
import kotlin.text.StringsKt;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.seimicrawler.xpath.JXDocument;
import org.seimicrawler.xpath.JXNode;

@Metadata(mv={1, 5, 1}, k=1, xi=48, d1={"\u0000&\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0004\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\b\n\u0002\u0018\u0002\n\u0002\b\u0002\u0018\u00002\u00020\u0001B\r\u0012\u0006\u0010\u0002\u001a\u00020\u0001\u00a2\u0006\u0002\u0010\u0003J\u001d\u0010\u0005\u001a\n\u0012\u0004\u0012\u00020\u0007\u0018\u00010\u00062\u0006\u0010\b\u001a\u00020\tH\u0000\u00a2\u0006\u0002\b\nJ\u0018\u0010\u000b\u001a\n\u0012\u0004\u0012\u00020\u0007\u0018\u00010\u00062\u0006\u0010\b\u001a\u00020\tH\u0002J\u0010\u0010\f\u001a\u0004\u0018\u00010\t2\u0006\u0010\r\u001a\u00020\tJ\u001b\u0010\u000e\u001a\b\u0012\u0004\u0012\u00020\t0\u00062\u0006\u0010\b\u001a\u00020\tH\u0000\u00a2\u0006\u0002\b\u000fJ\u0010\u0010\u0010\u001a\u00020\u00012\u0006\u0010\u0002\u001a\u00020\u0001H\u0002J\u0010\u0010\u0011\u001a\u00020\u00122\u0006\u0010\u0013\u001a\u00020\tH\u0002R\u000e\u0010\u0004\u001a\u00020\u0001X\u0082\u000e\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0014"}, d2={"Lio/legado/app/model/analyzeRule/AnalyzeByXPath;", "", "doc", "(Ljava/lang/Object;)V", "jxNode", "getElements", "", "Lorg/seimicrawler/xpath/JXNode;", "xPath", "", "getElements$reader_pro", "getResult", "getString", "rule", "getStringList", "getStringList$reader_pro", "parse", "strToJXDocument", "Lorg/seimicrawler/xpath/JXDocument;", "html", "reader-pro"})
public final class AnalyzeByXPath {
    @NotNull
    private Object jxNode;

    public AnalyzeByXPath(@NotNull Object doc) {
        Intrinsics.checkNotNullParameter((Object)doc, (String)"doc");
        this.jxNode = this.parse(doc);
    }

    private final Object parse(Object doc) {
        Object object;
        Object object2 = doc;
        if (object2 instanceof JXNode) {
            object = ((JXNode)doc).isElement() ? doc : this.strToJXDocument(doc.toString());
        } else if (object2 instanceof Document) {
            JXDocument jXDocument = JXDocument.create((Document)((Document)doc));
            Intrinsics.checkNotNullExpressionValue((Object)jXDocument, (String)"create(doc)");
            object = jXDocument;
        } else if (object2 instanceof Element) {
            Element[] elementArray = new Element[]{(Element)doc};
            JXDocument jXDocument = JXDocument.create((Elements)new Elements(elementArray));
            Intrinsics.checkNotNullExpressionValue((Object)jXDocument, (String)"create(Elements(doc))");
            object = jXDocument;
        } else if (object2 instanceof Elements) {
            JXDocument jXDocument = JXDocument.create((Elements)((Elements)doc));
            Intrinsics.checkNotNullExpressionValue((Object)jXDocument, (String)"create(doc)");
            object = jXDocument;
        } else {
            object = this.strToJXDocument(doc.toString());
        }
        return object;
    }

    private final JXDocument strToJXDocument(String html) {
        String html1 = html;
        if (StringsKt.endsWith$default((String)html1, (String)"</td>", (boolean)false, (int)2, null)) {
            html1 = "<tr>" + html1 + "</tr>";
        }
        if (StringsKt.endsWith$default((String)html1, (String)"</tr>", (boolean)false, (int)2, null) || StringsKt.endsWith$default((String)html1, (String)"</tbody>", (boolean)false, (int)2, null)) {
            html1 = "<table>" + html1 + "</table>";
        }
        JXDocument jXDocument = JXDocument.create((String)html1);
        Intrinsics.checkNotNullExpressionValue((Object)jXDocument, (String)"create(html1)");
        return jXDocument;
    }

    private final List<JXNode> getResult(String xPath) {
        Object node = this.jxNode;
        return node instanceof JXNode ? ((JXNode)node).sel(xPath) : ((JXDocument)node).selN(xPath);
    }

    @Nullable
    public final List<JXNode> getElements$reader_pro(@NotNull String xPath) {
        Intrinsics.checkNotNullParameter((Object)xPath, (String)"xPath");
        CharSequence charSequence = xPath;
        boolean bl = false;
        if (charSequence.length() == 0) {
            return null;
        }
        ArrayList jxNodes = new ArrayList();
        RuleAnalyzer ruleAnalyzes = new RuleAnalyzer(xPath, false, 2, null);
        Object object = new String[]{"&&", "||", "%%"};
        ArrayList<String> rules = ruleAnalyzes.splitRule((String)object);
        if (rules.size() == 1) {
            object = rules.get(0);
            Intrinsics.checkNotNullExpressionValue((Object)object, (String)"rules[0]");
            return this.getResult((String)object);
        }
        ArrayList<List<JXNode>> results = new ArrayList<List<JXNode>>();
        for (String string : rules) {
            Intrinsics.checkNotNullExpressionValue((Object)string, (String)"rl");
            List<JXNode> temp = this.getElements$reader_pro(string);
            if (temp == null) continue;
            Object object2 = temp;
            boolean bl2 = false;
            if (!(!object2.isEmpty())) continue;
            results.add(temp);
            object2 = temp;
            bl2 = false;
            if (!(!object2.isEmpty()) || !Intrinsics.areEqual((Object)ruleAnalyzes.getElementsType(), (Object)"||")) continue;
        }
        if (results.size() > 0) {
            if (Intrinsics.areEqual((Object)"%%", (Object)ruleAnalyzes.getElementsType())) {
                int n = 0;
                int n2 = ((List)results.get(0)).size() + -1;
                if (n <= n2) {
                    do {
                        int i = n++;
                        for (List list2 : results) {
                            if (i >= list2.size()) continue;
                            jxNodes.add(list2.get(i));
                        }
                    } while (n <= n2);
                }
            } else {
                for (List list3 : results) {
                    jxNodes.addAll(list3);
                }
            }
        }
        return jxNodes;
    }

    /*
     * WARNING - void declaration
     */
    @NotNull
    public final List<String> getStringList$reader_pro(@NotNull String xPath) {
        Intrinsics.checkNotNullParameter((Object)xPath, (String)"xPath");
        ArrayList<String> result2 = new ArrayList<String>();
        RuleAnalyzer ruleAnalyzes = new RuleAnalyzer(xPath, false, 2, null);
        Object object = new String[]{"&&", "||", "%%"};
        ArrayList<String> rules = ruleAnalyzes.splitRule((String)object);
        if (rules.size() == 1) {
            object = this.getResult(xPath);
            if (object != null) {
                void $this$mapTo$iv$iv;
                Iterable $this$map$iv = (Iterable)object;
                boolean bl = false;
                Iterable iterable = $this$map$iv;
                Collection destination$iv$iv = new ArrayList(CollectionsKt.collectionSizeOrDefault((Iterable)$this$map$iv, (int)10));
                boolean bl2 = false;
                for (Object item$iv$iv : $this$mapTo$iv$iv) {
                    void it;
                    JXNode jXNode = (JXNode)item$iv$iv;
                    Collection collection = destination$iv$iv;
                    boolean bl3 = false;
                    Boolean bl4 = result2.add(it.asString());
                    collection.add(bl4);
                }
                List cfr_ignored_0 = (List)destination$iv$iv;
            }
            return result2;
        }
        ArrayList<List<String>> results = new ArrayList<List<String>>();
        for (String string : rules) {
            Intrinsics.checkNotNullExpressionValue((Object)string, (String)"rl");
            List<String> temp = this.getStringList$reader_pro(string);
            Object object2 = temp;
            boolean bl = false;
            if (!(!object2.isEmpty())) continue;
            results.add(temp);
            object2 = temp;
            bl = false;
            if (!(!object2.isEmpty()) || !Intrinsics.areEqual((Object)ruleAnalyzes.getElementsType(), (Object)"||")) continue;
        }
        if (results.size() > 0) {
            if (Intrinsics.areEqual((Object)"%%", (Object)ruleAnalyzes.getElementsType())) {
                int n = 0;
                int n2 = ((List)results.get(0)).size() + -1;
                if (n <= n2) {
                    do {
                        int i = n++;
                        for (List list2 : results) {
                            if (i >= list2.size()) continue;
                            result2.add((String)list2.get(i));
                        }
                    } while (n <= n2);
                }
            } else {
                for (List list3 : results) {
                    result2.addAll(list3);
                }
            }
        }
        return result2;
    }

    @Nullable
    public final String getString(@NotNull String rule) {
        Intrinsics.checkNotNullParameter((Object)rule, (String)"rule");
        RuleAnalyzer ruleAnalyzes = new RuleAnalyzer(rule, false, 2, null);
        Object object = new String[]{"&&", "||"};
        ArrayList<String> rules = ruleAnalyzes.splitRule((String)object);
        if (rules.size() == 1) {
            object = this.getResult(rule);
            if (object != null) {
                Object object2 = object;
                boolean bl = false;
                boolean bl2 = false;
                Object it = object2;
                boolean bl3 = false;
                return TextUtils.join((CharSequence)"\n", (Iterable)it);
            }
            return null;
        }
        boolean bl = false;
        ArrayList<String> textList = new ArrayList<String>();
        for (String rl : rules) {
            Intrinsics.checkNotNullExpressionValue((Object)rl, (String)"rl");
            String temp = this.getString(rl);
            CharSequence charSequence = temp;
            boolean bl4 = false;
            boolean bl5 = false;
            if (charSequence == null || charSequence.length() == 0) continue;
            textList.add(temp);
            if (!Intrinsics.areEqual((Object)ruleAnalyzes.getElementsType(), (Object)"||")) continue;
            break;
        }
        return CollectionsKt.joinToString$default((Iterable)textList, (CharSequence)"\n", null, null, (int)0, null, null, (int)62, null);
    }
}

File: best-of-3/business/io/legado/app/model/analyzeRule/AnalyzeByXPath.kt
/** Business rewrite from reader-pro-3.2.14.jar — phase3. */

package io.legado.app.model.analyzeRule

import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import org.seimicrawler.xpath.JXDocument
import org.seimicrawler.xpath.JXNode

/**
 * XPath evaluator using seimicrawler (same as jar).
 * Supports && (union all) || (first non-empty) %% (zip by index) via RuleAnalyzer.
 */
class AnalyzeByXPath(doc: Any?) {
    private val jxNode: Any = parse(doc)

    private fun parse(doc: Any?): Any {
        return when (doc) {
            is JXNode -> if (doc.isElement) doc else strToJXDocument(doc.toString())
            is Document -> JXDocument.create(doc)
            is Element -> JXDocument.create(Elements(doc))
            is Elements -> JXDocument.create(doc)
            is String -> strToJXDocument(doc)
            null -> strToJXDocument("")
            else -> strToJXDocument(doc.toString())
        }
    }

    /** Wrap incomplete HTML fragments so XPath engine can parse tables. */
    private fun strToJXDocument(html: String): JXDocument {
        var html1 = html
        if (html1.endsWith("</td>")) html1 = "<tr>$html1</tr>"
        if (html1.endsWith("</tr>") || html1.endsWith("</tbody>")) html1 = "<table>$html1</table>"
        return JXDocument.create(html1)
    }

    private fun getResult(xPath: String): List<JXNode> {
        return try {
            when (val n = jxNode) {
                is JXNode -> n.sel(xPath) ?: emptyList()
                is JXDocument -> n.selN(xPath) ?: emptyList()
                else -> emptyList()
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun getElements(xPath: String): List<Any> {
        if (xPath.isEmpty()) return emptyList()
        val analyzer = RuleAnalyzer(xPath)
        val rules = analyzer.splitRule("&&", "||", "%%")
        if (rules.size == 1) return getResult(rules[0])

        val results = ArrayList<List<JXNode>>()
        for (rl in rules) {
            val part = getElements(rl).filterIsInstance<JXNode>()
            if (part.isNotEmpty()) {
                results.add(part)
                if (analyzer.elementsType == "||") break
            }
        }
        if (results.isEmpty()) return emptyList()
        return when (analyzer.elementsType) {
            "%%" -> zipByIndex(results)
            else -> results.flatten()
        }
    }

    fun getStringList(xPath: String): List<String> {
        if (xPath.isEmpty()) return emptyList()
        val analyzer = RuleAnalyzer(xPath)
        val rules = analyzer.splitRule("&&", "||", "%%")
        if (rules.size == 1) {
            return getResult(xPath).map { it.asString() ?: "" }
        }
        val results = ArrayList<List<String>>()
        for (rl in rules) {
            val part = getStringList(rl)
            if (part.isNotEmpty()) {
                results.add(part)
                if (analyzer.elementsType == "||") break
            }
        }
        if (results.isEmpty()) return emptyList()
        return when (analyzer.elementsType) {
            "%%" -> zipByIndex(results)
            else -> results.flatten()
        }
    }

    fun getString(rule: String): String {
        val analyzer = RuleAnalyzer(rule)
        val rules = analyzer.splitRule("&&", "||")
        if (rules.size == 1) {
            val nodes = getResult(rule)
            return nodes.joinToString("\n") { it.asString() ?: "" }
        }
        val parts = ArrayList<String>()
        for (rl in rules) {
            val s = getString(rl)
            if (s.isNotEmpty()) {
                parts += s
                if (analyzer.elementsType == "||") break
            }
        }
        return parts.joinToString("\n")
    }

    // Adapt to AnalyzeRule call sites that pass (content, rule)
    fun getString(content: Any?, rule: String): String = AnalyzeByXPath(content ?: jxNode).getString(rule)
    fun getStringList(content: Any?, rule: String): List<String> = AnalyzeByXPath(content ?: jxNode).getStringList(rule)
    fun getElements(content: Any?, rule: String): List<Any> = AnalyzeByXPath(content ?: jxNode).getElements(rule)

    private fun <T> zipByIndex(results: List<List<T>>): List<T> {
        if (results.isEmpty()) return emptyList()
        val out = ArrayList<T>()
        val max = results.maxOf { it.size }
        for (i in 0 until max) {
            for (list in results) {
                if (i < list.size) out += list[i]
            }
        }
        return out
    }
}
